### DISCUSSION TITLE
Optimization for gigantic amount of nodes

### DISCUSSION THREAD

**Discussion by MunifTanjim** [nvim-neo-tree/neo-tree.nvim#995] [Internal Development]
I'm trying out a Linked List based internal for NuiTree.

- https://github.com/MunifTanjim/nui.nvim/pull/285

Currently `tree:render()` recalculates everything each time it is called. That means:
- Iterating over all the nodes to prepare the display lines
- Update the whole buffer with the newly prepared display lines

For projects with a huge number of files/folders, it becomes slow as more and more folders are expanded.

With the new implementation the goal is to:
- Be able to partially calculate display lines for the changed nodes
- Be able to update the buffer partially

I tried out `neo-tree` with the new NuiTree internal. But I'm facing some weird bugs.

This discussion is to figure out what's wrong...

Does `neo-tree` mutate any internals of NuiTree that might be causing the issue? If yes, what new methods should NuiTree expose so that `neo-tree` does not need to hack into NuiTree internals anymore?

_P.S.: https://github.com/MunifTanjim/nui.nvim/pull/285 is not usable yet._

**Comment by MunifTanjim** [comment]
cc: @cseickel (not sure who else is maintaining the plugin)

**Reply by cseickel** [reply]
@nhat-vo @miversen33 @pysan3 are the active maintainers.

**Comment by cseickel** [comment]
> I'm trying out a Linked List based internal for NuiTree.
> ...
> With the new implementation the goal is to:
> 
> * Be able to partially calculate display lines for the changed nodes
> * Be able to update the buffer partially

That's awesome! 

> I tried out `neo-tree` with the new NuiTree internal. But I'm facing some weird bugs.
> 
> This discussion is to figure out what's wrong...
> 
> Does `neo-tree` mutate any internals of NuiTree that might be causing the issue? If yes, what new methods should NuiTree expose so that `neo-tree` does not need to hack into NuiTree internals anymore?

If we do, it would be within this file: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/v2.x/lua/neo-tree/ui/renderer.lua

Could you describe some of the bugs? I might know where to look if I understand a little more about what breaks.

**Reply by MunifTanjim** [reply]
> Could you describe some of the bugs?

With the latest changes on https://github.com/MunifTanjim/nui.nvim/pull/285, when I run `"refresh"` I see a glitch. 

![Kapture 2023-06-17 at 20 12 40](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/8050659/daff31ab-0efb-4b10-a276-6e1e0219a2bd)

Looks like many nodes are rendered on top of the tree and then it fixes itself automatically.

The same thing is happening when I add/remove a new file.

Not really sure what's causing this. Most probably this is a bug in the new NuiTree implementation. I'm just trying to figure out how to isolate the bug so that I can write a test-case for it.

**Reply by MunifTanjim** [reply]
Looks like whenever I'm adding/removing a file, `tree:render()` is being called 5 times 🤔

**Reply by MunifTanjim** [reply]
And for "refresh", `tree:render()` is being called 2 times.

**Reply by MunifTanjim** [reply]
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/d883632bf8f92f1d5abea4a9c28fb2f90aa795aa/lua/neo-tree/ui/renderer.lua#L1194-L1219

What does this do btw? 🤔

Looks like this is changing one of the nodes and then calling `tree:set_nodes()`? Is it one of the hacks? If the intention is to only change a single node, I can expose a method for doing that.

**Reply by cseickel** [reply]
The last item is for grouping folders that only contain a single folder, so if you have this structure:

- org/
  - github/
     - actions/
       - file
       - file

it would become:

- org/github/actions/
  - file
  - file  


One reason for multiple renders is that it will scan the filesystem and query git separately and then display the files and update them with git status in a second render whenever that is ready. When adding or removing files, I can see further renders happening because of lsp diagnostics reporting updates.

There may or may not be  bug in neo-tree that causes extra renders, but either way I think that is something Nui should be able to handle. Maybe new renders are coming before the old one finishes and they are clobbering each other. I would say there should be an internal lock to delay new render requests when one is under way. I have an internal utility for debouncing things like that that you can use if you want:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/d883632bf8f92f1d5abea4a9c28fb2f90aa795aa/lua/neo-tree/utils.lua#L68-L73

**Reply by MunifTanjim** [reply]
Phew... so I made some changes and now it's works without any changes to `neo-tree`.

😂 Now gotta do some benchmark to see if it actually improves performance or not.

**Reply by cseickel** [reply]
It's a showdown between lua and neovim core. I'm betting the new code wins in a large tree with small changes. 

I'd guess that the majority of renders in neo-tree only change 1 character.

**Reply by miversen33** [reply]
This actually raises an interesting point.
@MunifTanjim , I haven't dug into the internals of Nui in quite some time. What are your thoughts on "per leaf" rendering as opposed to entire tree rendering?

I don't know how much effort that would be, but from purely a performance mindset, I imagine that a majority of tree's don't completely change with each render. A new branch/leaf might be added or pruned, but the overall tree will remain mostly the same.

**Reply by MunifTanjim** [reply]
> What are your thoughts on "per leaf" rendering as opposed to entire tree rendering?

That is what the open PR in doing. I haven't  benchmarked it yet. Once that is done, I'll merge that. 

@miversen33

**Reply by miversen33** [reply]
I suspected that is what this was doing but had a tough time being completely sure :) Thanks for clarifying!
